docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,120 @@
+* We currently do not have a way to act on failed/crashed NBD volumes with the existing healer logic. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it.
+* This will be a generic framework for any userspace mounters added in the future, example ceph-fuse
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller driver. It is the one who triggers controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
+    * Fetch all the volume attachments (VA) list
+    * Filter the volume attachments list through matching driver name and status attached
+    * For each volume attachment get the respective PV information and check the criteria of PV Bound (also we can have configuration options like mounter type)
+    * Send the RPC request for side-car to check if the volume is healthy along with the PV details details
+    * Log the final response returned from the side-car
+* The healer controller periodically triggers the health status checks of the CSI volumes via the gRPC calls.
+
+## CSI-Addons side-car
+* The side-car registers itself by creating a `CSIAddonsNode` CR that the CSI-Addons Controller can use to connect to the side-car and execute operations.
+* When a nodeplugin container is restarted the side-car will update/create(if not present) [CSIAddonsNode](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/api/v1alpha1/csiaddonsnode_types.go) object.
+* The side-car works as relay, upon receiving a gRPC call from healer controller, the CSI-Addons side-car will then check volume's mounting conditions collected via the CSI NodeGetVolumeStats RPC.
+* In case if any volume condition is reported abnormal, the CSI-Addons side-car will act on it, it does some internal operation like fetching secrets, constructing staging path and finally sends NodeStageVolume request to the CSI driver and finally returns back the response to the controller.
+
+PS: Some modification might be required in side-car to update the `CSIAddonsNode` object if its already present and its minimal change.
```

Comment:
not sure what you mean by this?

Reply:
Soon after the sidecar is restarted, if it sees `CSIAddonsNode` exist, we need to implement a way to update the already existing `CSIAddonsNode` object.

Reply:
that should be handled by the CSIAddonsNode feature, it is not relevant enough to mention here

I think you are concerned that the CSI-driver container in the pod restarted, and not the pod was re-created. Only in that case, the CSIAddonsNode object would exist. It is not clear to me if this is something that can happen in reality.